### PR TITLE
GrafanaUI: Fix styles for invalid selects & DataSourcePicker

### LINKED ISSUE #52526: Select: Select with invalid property set to true doesn't change border color
It seems that setting `invalid={true}` prop on Select doesn't change the border color to read. 
You can reproduce it also in our Storybook https://developers.grafana.com/ui/latest/index.html?path=/story/forms-select--basic&args=invalid:true. 
**What happened**:

**What you expected to happen**:
If I set `invalid={true}`, the border of the select should be read.

**How to reproduce it (as minimally and precisely as possible)**:

Set `invalid={true}` in Datasource picker in Explore ([specifically here](https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/components/DataSourcePicker.tsx#L189)), the border doesn't turn read. 

**Anything else we need to know?**:

**Environment**:
- Grafana version: main/v9.0+

### PR DESCRIPTION
**Pull request by Elfo404**
Selects (and by consequence DataSourcePicker) were not following other inputs styles when they were invalid as shown in the screenshots (or in https://developers.grafana.com/ui/canary/index.html?path=/story/forms-select--basic&args=invalid:true).

this PR changes the styles for the BaseSelect component to that it add the red border like other inputs, plus add an `invalid` prop to the DatasourcePicker component to achieve the same thing (the prop is then passed down to the Select component.

Before (invalid=true):
Screenshot 2022-08-09 at 16 36 24

After:
Screenshot 2022-08-09 at 16 37 26

### CODE REVIEW

**Review by eledobleefe — COMMENTED** [review]
Hi @Elfo404! Looking good :)
Just one comment. There is a duplicate of SelectContainer.tsx. It's the file named Container.tsx (packages/grafana-ui/src/components/Select/Container.tsx), would you mind removing it in your PR?
Thank you for all your work! :)

**Review by leventebalogh — APPROVED** [review]
LGTM 👍

File: packages/grafana-runtime/src/components/DataSourcePicker.tsx [hunk 1]
```diff
@@ -186,7 +187,7 @@ export class DataSourcePicker extends PureComponent<DataSourcePickerProps, DataS
           placeholder={placeholder}
           noOptionsMessage="No datasources found"
           value={value ?? null}
-          invalid={!!error}
+          invalid={!!error || !!this.props.invalid}
```
Comment (dprokop) [inline packages/grafana-runtime/src/components/DataSourcePicker.tsx:190]: Readability nit: 
```suggestion
          invalid={Boolean(error) || Boolean(this.props.invalid)}
```

File: packages/grafana-ui/src/components/Select/SelectBase.tsx [hunk 2]
```diff
@@ -317,28 +317,30 @@ export function SelectBase<T>({
...
           },
-          DropdownIndicator(props: any) {
+          DropdownIndicator(props: DropdownIndicatorProps) {
             return <DropdownIndicator isOpen={props.selectProps.menuIsOpen} />;
           },
           SingleValue(props: any) {
             return <SingleValue {...props} disabled={disabled} />;
           },
+          SelectContainer(props: ContainerProps) {
+            return <SelectContainer {...props} invalid={!!invalid} />;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Select/SelectBase.tsx:340]:
```suggestion
            return <SelectContainer {...props} invalid={Boolean(invalid)} />;
```

**Review by dprokop — APPROVED** [review]

**Review by Clarity-89 — COMMENTED** [review]
Looks good, but it seems like lots of the tests are failing. Wonder if it's because our `Field` and `InlineField` components pass down own `invalid` prop, which would override Select's `invalid` prop? E.g.: https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Forms/InlineField.tsx#L72

**Comment by Elfo404** [comment]
@dprokop @leventebalogh @joshhunt 
can you please retake a look at this? it's already approved but i did some changes to fix a couple of tests. 
also i'm not really comfortable with this change:
https://github.com/grafana/grafana/pull/53476/files#diff-9e7b4287b9a4e1bd881de03e6ad6be26a5a4850fc1553b4a33a50c6d7c66e088R26-R27

but i can't get my way around it.

**Comment by joshhunt** [comment]
@Elfo404 Yeah I'll take a look at this next week to sort it out!!

**Comment by joshhunt** [comment]
@eledobleefe and I have worked on fixing the types for the custom component props:

 - `invalid` is not a prop of react-select, but we pass it in anyway, so it's not on the default `selectProps` type that custom components use
 - we created a `ReactSelectProps` type that extends the regular `selectProps` and adds on some of the custom props we pass into react-select, and then use that with `CustomComponentProps`, which is what custom controls should use
   - Note: the props we pass into react-select (including the custom ones) do not use the `ReactSelectProps` type, so there's little type safety when using this at the moment

@Elfo404 Would you mind checking this over and making sure it works with datasource picker and explore please? 🙏🏻

**Comment by Elfo404** [comment]
@joshhunt like a charm, thank you for looking at this! ❤️ 
Screenshot 2022-08-26 at 11 32 08
Screenshot 2022-08-26 at 11 33 36

**Pull request merged**